tiledb/array.py

```diff
@@ -19,20 +18,21 @@ class DenseArray(DenseArrayImpl):
     and assignment.
     """
 
-    _mixin_init = False
-
     def __new__(cls, *args, **kwargs):
-        if not cls._mixin_init:
-            # must set before importing, because import is not thread-safe
-            #   https://github.com/TileDB-Inc/TileDB-Py/issues/244
-            cls._mixin_init = True
-            try:
-                from tiledb.cloud import cloudarray
-
-                DenseArray.__bases__ = DenseArray.__bases__ + (cloudarray.CloudArray,)
-            except ImportError:
-                pass
-
+        try:
+            from tiledb.cloud import cloudarray
+
+            cls.__bases__ += (cloudarray.CloudArray,)
+        except (ImportError, TypeError):
+            # If we couldn't import, tiledb.cloud isn't installed
+            # If there was a TypeError, CloudArray was already added (by another thread)
+            pass
+        try:
+            # Mixin initialization completed: delete this method for subsequent calls
+            del DenseArray.__new__
```

Comment:
I am admittedly a little wary of modifying object pieces like this in the middle of a call (e.g. is `super()` guaranteed to handle the disappearance of `__new__` properly if it’s deleted as it runs?) but I can confirm that, in the absence of thread concerns, this works:

```
In [1]: class A:
   ...:     def __new__(cls, *args, **kwargs):
   ...:         print(f"A.__new__({cls}, ...)")
   ...:         ret = super().__new__(cls, *args, **kwargs)
   ...:         print("A.__new__ done")
   ...: 

In [2]: class B(A):
   ...:     def __new__(cls, *args, **kwargs):
   ...:         print(f"B.__new__({cls}, ...")
   ...:         ret = super().__new__(cls, *args, **kwargs)
   ...:         print("B.__new__ done")
   ...: 

In [3]: a = A()
A.__new__(<class '__main__.A'>, ...)
A.__new__ done

In [4]: b = B()
B.__new__(<class '__main__.B'>, ...
A.__new__(<class '__main__.B'>, ...)
A.__new__ done
B.__new__ done

In [5]: del B.__new__

In [6]: b2 = B()
A.__new__(<class '__main__.B'>, ...)
A.__new__ done
```

Reply:
ha ha ha i forgot to `return ret` good job me

Reply:
As commented above, in general this would need more testing in the presence of (single, let alone multiple) inheritance. In this case we can even replace `super()` with `DenseArrayImpl/SparseArrayImpl` and not worry about method resolution.